language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -421,10 +530,155 @@ genModuleRef (pkgRef, modName) = case pkgRef of
+  (["{"
+   , "  \"private\": true,"
+   , "  \"name\": \"" <> packageName <> "\","
+   , "  \"version\": \"" <> packageVersion <> "\","
+   , "  \"description\": \"Produced by daml2ts\","
+   , "  \"license\": \"Apache-2.0\","
+   , "  \"dependencies\": {"
+   , "    \"@daml/types\": \"" <> damlTypesVersion <> "\","
+   , "    \"@mojotech/json-type-validation\": \"^3.1.0\"" ++ if not $ null dependencies then ", " else ""
+   ] ++ dependencies ++
+    ["  },"
+    , "  \"scripts\": {"
+    , "    \"build\": \"tsc --build\","
+    , "    \"lint\": \"eslint --ext .ts src/ --max-warnings 0\""
+    , "  },"
+    , "  \"devDependencies\": {"
+    , "    \"@typescript-eslint/eslint-plugin\": \"^2.11.0\","
+    , "    \"@typescript-eslint/parser\": \"^2.11.0\","
+    , "    \"eslint\": \"^6.7.2\","
+    , "    \"typescript\": \"~3.7.3\""
+    , "  }"
+    , "}"
+    ])
+  where
+    packageName =  packageNameOfPackageDir packageDir
+    dependencies = withCommas [ "    \"" <> pkg <> "\": \"" <> packageVersion <> "\""
+                              | d <- depends
+                              , let pkg = scope ++ "/" ++ undependency d
+                              ]
+
+    withCommas :: [String] -> [String]
+    withCommas [] = []
+    withCommas ms = reverse (head ms' : map (++",") (tail ms')) where ms' = reverse ms
+
+-- This type describes the format of a "top-level" 'package.json'. We
+-- expect such files to have the format
+-- {
+--   "private": true,
+--   "workspaces: [
+--      "path/to/foo",
+--      "path/to/bar",
+--      ...
+--   ]
+-- }
+data PackageJson = PackageJson
+  {  private :: Bool
+  ,  workspaces :: [T.Text]
+  } deriving Show
+-- Explicitly provide instances to avoid relying on restricted
+-- extension 'DeriveGeneric'.
+instance FromJSON PackageJson where
+  parseJSON (Object v) = PackageJson
+      <$> v .: "private"
+      <*> v .: "workspaces"
+  parseJSON _ = mzero
+instance ToJSON PackageJson where
+  toJSON (PackageJson private workspaces) =
+    object [ "private" .= private
+           , "workspaces" .= workspaces
+           ]
+
+-- Read the provided 'package.json'; transform it to include the
+-- provided workspaces; write it back to disk.
+writeTopLevelPackageJson :: FilePath -> [(String, String, [Dependency])] -> FilePath -> IO ()
+writeTopLevelPackageJson optOutputDir dependencies file = do
+  let (g, nodeFromVertex) = graphFromEdges'
+        (map (\(a, b, ds) -> (a, b, map undependency ds)) dependencies)
+      ps = map (fst3 . nodeFromVertex) $ reverse (topSort g)
+        -- Topologically order our packages.
+      ldr = fromJust (stripPrefix "@" (unscope (scopeOfScopeDir optOutputDir)))
+        -- 'ldr' we expect to be something like "daml2ts/".
+  let ourPackages = map (T.pack . ((ldr ++ "/") ++)) ps
+  bytes <- BSL.readFile file
+  maybe
+    (fail $ "Error decoding JSON from '" <> file <> "'")
+    (transformAndWrite ourPackages ldr)
+    (decode bytes :: Maybe PackageJson)
```

Comment:
This looks somewhat dangerous. You are throwing away every field apart from the two you modify. Additional fields should probably stay intact. Let’s tackle that in a separate PR.

Reply:
sounds good. yes, i am a newbie with this library and would like to learn how to better here in this respect.

Reply:
@hurryabit addresses this (along with other niggles) in https://github.com/digital-asset/daml/pull/4820.

Reply:
This is fixed in https://github.com/digital-asset/daml/pull/4820.